src/expression.rs

```diff
@@ -194,62 +194,74 @@ impl Expression {
     /// use std::str::FromStr;
     /// use num_complex::Complex64;
     ///
-    /// let expression = Expression::from_str("cos(2 * pi) + 2").unwrap().simplify();
+    /// let mut expression = Expression::from_str("cos(2 * pi) + 2").unwrap();
+    /// expression.simplify();
     ///
     /// assert_eq!(expression, Expression::Number(Complex64::from(3.0)));
     /// ```
-    pub fn simplify(self) -> Self {
+    pub fn simplify(&mut self) {
         use Expression::*;
 
-        let simplified = match self {
+        match self {
             FunctionCall {
                 function,
                 expression,
             } => {
-                let simplified = expression.simplify();
-                if let Number(number) = simplified {
-                    Number(calculate_function(&function, &number))
-                } else {
-                    FunctionCall {
-                        function,
-                        expression: Box::new(simplified),
-                    }
+                expression.simplify();
+                if let Number(number) = expression.as_ref() {
+                    *self = Number(calculate_function(function, number));
                 }
             }
             Infix {
                 left,
-                operator,
+                operator: _,
                 right,
-            } => Infix {
-                left: Box::new(left.simplify()),
-                operator,
-                right: Box::new(right.simplify()),
-            },
+            } => {
+                left.simplify();
+                right.simplify();
```

Comment:
Might we want to handle a case where the operands simplify to constants and this can be completely evaluated?

Reply:
Good thought - but there's a second step to evaluation below that will collapse both this and `Prefix` (unless you see a bug/case that I don't)

Although - this is inconsistent with the`FunctionCall` branch, which takes it upon itself to evaluate inline. We should go one way or the other.